appserver/payara-appserver-modules/microprofile/fault-tolerance/src/main/java/fish/payara/microprofile/faulttolerance/service/FaultToleranceServiceImpl.java

```diff
@@ -148,112 +171,63 @@ public void event(Event<?> event) {
-
+        for (Entry<String, FaultToleranceMethodContextImpl> methodValue : methodByTargetObjectAndName.entrySet()) {
+            String group = methodValue.getKey();
+            MonitoringDataCollector methodCollector = collector.group(group);
+            FaultToleranceMethodContext context = methodValue.getValue();
+            BlockingQueue<Thread> concurrentExecutions = context.getConcurrentExecutions(-1);
+            if (concurrentExecutions != null) {
+                collectBulkheadSemaphores(methodCollector, concurrentExecutions);
+                collectBulkheadSemaphores(methodCollector, concurrentExecutions, context.getQueuingOrRunningPopulation());
+            }
+            collectCircuitBreakerState(methodCollector, context.getState(-1));
         }
     }
 
-    private static <V> void collectMethodState(MonitoringDataCollector collector, String appName, String type,
-            Map<Object, Map<String, V>> entries, BiConsumer<MonitoringDataCollector, V> collect) {
-        for (Entry<Object, Map<String, V>> entry : entries.entrySet()) {
-            Object target = entry.getKey();
-            String targetValue = System.identityHashCode(target) + "@" + target.getClass().getSimpleName();
-            for (Entry<String, V> methodValue : entry.getValue().entrySet()) {
-                String group = appName + "-" + type + "-" + targetValue + "-" + methodValue.getKey();
-                collect.accept(collector.group(group), methodValue.getValue());
-            }
-        }
+    private static void collectBulkheadSemaphores(MonitoringDataCollector collector,
+            BlockingQueue<Thread> concurrentExecutions) {
+        collector
+            .collect("RemainingConcurrentExecutionsCapacity", concurrentExecutions.remainingCapacity())
+            .collect("ConcurrentExecutions", concurrentExecutions.size());
     }
 
-    private static void collectBulkheadSemaphores(MonitoringDataCollector collector, BulkheadSemaphore semaphore) {
+    private static void collectBulkheadSemaphores(MonitoringDataCollector collector,
+            BlockingQueue<Thread> concurrentExecutions, AtomicInteger queuingOrRunningPopulation) {
         collector
-            .collect("availablePermits", semaphore.availablePermits())
-            .collect("acquiredPermits", semaphore.acquiredPermits());
+            .collect("WaitingQueuePopulation", queuingOrRunningPopulation.get() - concurrentExecutions.size());
     }
 
     private static void collectCircuitBreakerState(MonitoringDataCollector collector, CircuitBreakerState state) {
+        if (state == null) {
+            return;
+        }
         collector
-            .collect("halfOpenSuccessFul", state.getHalfOpenSuccessFulResultCounter())
-            .collect("state", state.getCircuitState().name().charAt(0));
+            .collect("circuitBreakerHalfOpenSuccessFul", state.getHalfOpenSuccessFulResultCounter())
```

Comment:
Only just noticed this since apparently it's already there in the CircuitBreakerState for _some_ method names: _SuccessFul_ shouldn't have a capital 'F'

o((>ω< ))o

Reply:
Going to fix that.